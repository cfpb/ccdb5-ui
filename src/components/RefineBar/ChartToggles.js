import './ChartToggles.less';
import { changeChartType } from '../../actions/trends';
import getIcon from '../iconMap';
import React from 'react';
import { selectedClass, sendAnalyticsEvent } from '../../utils';
import { useDispatch, useSelector } from 'react-redux';
import { selectTrendsChartType } from '../../reducers/trends/selectors';

export const ChartToggles = () => {
  const dispatch = useDispatch();
  const chartType = useSelector(selectTrendsChartType);

  const toggleChartType = (chartType) => {
    sendAnalyticsEvent('Button', 'Trends:' + chartType);
    dispatch(changeChartType(chartType));
  };

  return (
    <section className="chart-toggles m-btn-group">
      <p>Chart type</p>
      <button
        aria-label="Toggle line chart"
        className={'a-btn' + selectedClass('line', chartType)}
        disabled={chartType === 'line'}
        onClick={() => {
          toggleChartType('line');
        }}
      >
        {getIcon('line-chart')}
      </button>
      <button
        aria-label="Toggle area chart"
        className={'a-btn' + selectedClass('area', chartType)}
        disabled={chartType === 'area'}
        onClick={() => {
          toggleChartType('area');
        }}
      >
        {getIcon('area-chart')}
      </button>
    </section>
  );
};
