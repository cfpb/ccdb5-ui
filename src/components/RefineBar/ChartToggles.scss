.chart-toggles {
  button {
    width: 38px;
    background-color: var(--pacific-40);
    padding: 3px;
    svg {
      height: 25px;

      &#line-chart-icon {
        .cls-1 {
          opacity: 0.4;
        }

        .cls-2 {
          fill: none;
          stroke: #101820;
          stroke-miterlimit: 10;
          stroke-width: 2px;
        }

        .cls-3 {
          fill: #101820;
        }
      }
      &#area-chart-icon {
        .cls-1 {
          fill: #101820;
        }

        .cls-2 {
          opacity: 0.4;
        }
      }
    }
  }
}
