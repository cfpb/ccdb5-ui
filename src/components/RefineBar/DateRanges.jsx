import './DateRanges.less';
import { connect } from 'react-redux';
import { dateRanges } from '../../constants';
import { dateRangeToggled } from '../../actions/filter';
import React from 'react';
import { sendAnalyticsEvent } from '../../utils';

export class DateRanges extends React.Component {
  _btnClassName( dateRange ) {
    const classes = [ 'a-btn', 'date-selector', 'range-' + dateRange ];
    if ( dateRange === this.props.dateRange ) {
      classes.push( 'selected' );
    }
    return classes.join( ' ' );
  }

  _toggleDateRange( dateRange ) {
    if ( this.props.dateRange !== dateRange ) {
      this.props.toggleDateRange( dateRange, this.props.tab );
    }
  }

  render() {
    return (
      <section className="date-ranges m-btn-group">
        <p>Date range (Click to modify range)</p>
        { dateRanges.map( dateRange => <button onClick={ () => {
          this._toggleDateRange( dateRange );
        } }
        className={ this._btnClassName( dateRange ) }
        key={ dateRange }>
          { dateRange }
        </button>
        ) }
      </section>
    );
  }
}

export const mapStateToProps = state => ( {
  dateRange: state.query.dateRange,
  tab: state.query.tab
} );

export const mapDispatchToProps = dispatch => ( {
  toggleDateRange: ( range, tab ) => {
    sendAnalyticsEvent( 'Button', tab + ':' + range );
    dispatch( dateRangeToggled( range ) );
  }
} );

export default connect( mapStateToProps, mapDispatchToProps )( DateRanges );
