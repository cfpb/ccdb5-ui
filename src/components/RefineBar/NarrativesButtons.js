import {
  filterAdded,
  filterRemoved,
} from '../../reducers/filters/filtersSlice';
import { useDispatch, useSelector } from 'react-redux';
import { selectedClass } from '../../utils';
import { selectFiltersHasNarrative } from '../../reducers/filters/selectors';

const FIELD_NAME = 'has_narrative';

export const NarrativesButtons = () => {
  const dispatch = useDispatch();
  const hasNarrative = useSelector(selectFiltersHasNarrative);
  const isNarrativesButtonDisabled = hasNarrative === true;

  return (
    <section className="narratives-buttons">
      <p>Read</p>
      <div className="m-btn-group">
        <button
          id="btn-add-narratives"
          className={'a-btn' + selectedClass(true, isNarrativesButtonDisabled)}
          disabled={isNarrativesButtonDisabled}
          onClick={() => {
            dispatch(filterAdded(FIELD_NAME, ''));
          }}
        >
          Only complaints with narratives
        </button>

        <button
          id="btn-remove-narratives"
          className={
            'a-btn' + selectedClass(false, !!isNarrativesButtonDisabled)
          }
          disabled={!isNarrativesButtonDisabled}
          onClick={() => {
            dispatch(filterRemoved(FIELD_NAME, ''));
          }}
        >
          All complaints
        </button>
      </div>
    </section>
  );
};
