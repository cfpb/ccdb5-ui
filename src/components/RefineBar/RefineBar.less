@import (less) "../../css/base.less";

.refine-bar {
  section {
    margin: @grid_gutter-width/3 0 @grid_gutter-width/3 @grid_gutter-width/2;
    p {
      font-size: @size-vi;
      font-weight: 600;
      color: @gray;
    }
    button {
      &.selected {
        font-weight: 600;
      }
    }
    &:not(.filter-panel-toggle){
      button {
        color: @black;
      }
    }

    &.cf-select {
      border: none;
      select {
        background-color: @white;
        color: @black;
        border: 1px solid @gray-40;
      }
      &::after {
        background-color: @white;
        transform: translate(-5px, 0px);
      }
    }
  }

  .separator {
    @media @desktop, @tablet-lg, @phone {
      display: none;
    }
  }

  // Per Capita and Narratives buttons in the refine bar.
  .toggle-button {
    // Only used to hide "population" on per capita refine buttons.
    @media @phone, @tablet-sm {
      > span {
        display: none;
      }
    }

    &.selected {
      pointer-events: none;
    }

    &:not(.a-btn__disabled) {
      background-color: @pacific-80;
    }
    &.deselected {
      background-color: @pacific-40;
    }
  }
}
