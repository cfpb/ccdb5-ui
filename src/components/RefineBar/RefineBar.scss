@use 'sass:math';
@use '@cfpb/cfpb-design-system/src/abstracts' as *;
@use 'css/base' as *;
@use 'select' as *;

.refine-bar {
  section {
    margin: math.div($grid-gutter-width, 3) 0 math.div($grid-gutter-width, 3) math.div($grid-gutter-width, 2);
    p {
      font-size: $size-vi;
      font-weight: 600;
      color: var(--gray);
    }
    button {
      &.selected {
        background-color: var(--pacific-60);
        font-weight: 600;
      }
      &:not(.selected) {
        background-color: var(--pacific-20);
      }
    }
    &:not(.filter-panel-toggle) {
      button {
        color: var(--black);
      }
    }

    &.cf-select {
      border: none;
      select {
        background-color: var(--white);
        color: var(--black);
        border: 1px solid var(--gray-40);
      }
      &::after {
        background-color: var(--white);
        transform: translate(-5px, 0);
      }
    }
  }

  .separator {
    @media screen and (min-width: $layout-breakpoint-desktop-min) {
      display: none;
    }

    @media screen and (min-width: $layout-breakpoint-tablet-lg-min) and (max-width: $layout-breakpoint-tablet-lg-max) {
      display: none;
    }

    @media screen and (max-width: $layout-breakpoint-phone-max) {
      display: none;
    }
  }

  // Per Capita and Narratives buttons in the refine bar.
  button {
    // Only used to hide "population" on per capita refine buttons.
    @media screen and (max-width: $layout-breakpoint-phone-max),
      screen and (min-width: $layout-breakpoint-tablet-sm-min) and (max-width: $layout-breakpoint-tablet-sm-max){
      > span {
        display: none;
      }
    }

    &.disabled.a-btn__disabled {
      background-color: var(--gray-10);
    }
  }
}
