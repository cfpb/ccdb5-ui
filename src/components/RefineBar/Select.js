import PropTypes from 'prop-types';
import { useMemo } from 'react';

export const Select = ({ id, handleChange, label, title, value, values }) => {
  const idSelect = 'select-' + id;
  const vals = useMemo(() => {
    // different cases that values can me:
    // Array
    // handle cases where an array of single entries
    // case 1: values = [1,2,4]
    // case 2: values = [
    // { name: 'Foo', disabled: false},
    // { name:'bar', disabled: true }
    // ]
    // object key val pair
    // case 3: values = {
    //   created_date_desc: 'Newest to oldest',
    //   created_date_asc: 'Oldest to newest',
    //   relevance_desc: 'Relevance',
    //   relevance_asc: 'Relevance (asc)'
    // }
    // array of objects

    if (Array.isArray(values)) {
      // do nothing, case 2
      if (Object.prototype.hasOwnProperty.call(values[0], 'name')) {
        return values;
      } else {
        // case 1
        return values.map((val) => ({
          name: val,
          value: val,
          disabled: val.disabled,
        }));
      }
    } else {
      // case 3
      return Object.keys(values).map((obj) => ({
        name: values[obj],
        value: obj,
        disabled: obj.disabled,
      }));
    }
  }, [values]);

  return (
    <section className="cf-select" data-tour={idSelect}>
      <label className="u-visually-hidden" htmlFor={idSelect}>
        {label}
      </label>
      <p>{title}</p>
      <select value={value} id={idSelect} onChange={handleChange}>
        {vals.map((val) => (
          <option
            disabled={[val.value, val.name].includes(value) || val.disabled}
            key={val.name}
            value={val.value || val.name}
          >
            {val.name}
          </option>
        ))}
      </select>
    </section>
  );
};

Select.propTypes = {
  id: PropTypes.string,
  handleChange: PropTypes.func.isRequired,
  label: PropTypes.string,
  title: PropTypes.string,
  values: PropTypes.oneOfType([PropTypes.array, PropTypes.object]).isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};
