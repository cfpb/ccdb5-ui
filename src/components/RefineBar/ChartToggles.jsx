import './ChartToggles.scss';
import { chartTypeUpdated } from '../../reducers/trends/trendsSlice';
import getIcon from '../Common/Icon/iconMap';
import { selectedClass, sendAnalyticsEvent } from '../../utils';
import { useDispatch, useSelector } from 'react-redux';
import { selectTrendsChartType } from '../../reducers/trends/selectors';

export const ChartToggles = () => {
  const dispatch = useDispatch();
  const chartType = useSelector(selectTrendsChartType);

  const toggleChartType = (chartType) => {
    sendAnalyticsEvent('Button', 'Trends:' + chartType);
    dispatch(chartTypeUpdated(chartType));
  };

  return (
    <section className="chart-toggles">
      <p>Chart type</p>
      <div className="m-btn-group">
        <button
          aria-label="Toggle line chart"
          className={'a-btn' + selectedClass('line', chartType)}
          disabled={chartType === 'line'}
          onClick={() => {
            toggleChartType('line');
          }}
        >
          {getIcon('line-chart')}
        </button>
        <button
          aria-label="Toggle area chart"
          className={'a-btn' + selectedClass('area', chartType)}
          disabled={chartType === 'area'}
          onClick={() => {
            toggleChartType('area');
          }}
        >
          {getIcon('area-chart')}
        </button>
      </div>
    </section>
  );
};
