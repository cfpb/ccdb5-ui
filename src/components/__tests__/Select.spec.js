import React from 'react';
import renderer from 'react-test-renderer';
import { Select } from '../RefineBar/Select';

describe('component:Select', () => {
  it('renders array values without crashing', () => {
    const options = ['Uno', 'Dos', 'Tres'];
    const target = renderer.create(
      <Select
        label="Select something"
        title="Show sumthing"
        values={options}
        id="txt"
        value="Dos"
        handleChange={jest.fn()}
      />
    );

    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders object values without crashing', () => {
    const options = {
      10: '10 results',
      25: '25 results',
      50: '50 results',
      100: '100 results',
    };
    const target = renderer.create(
      <Select
        label="Select size"
        title="Show"
        values={options}
        id="size"
        value={10}
        handleChange={jest.fn()}
      />
    );

    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders array of objects without crashing', () => {
    const options = [
      { name: 'Uno', enabled: true },
      { name: 'Dos', enabled: false },
      { name: 'Tres', enabled: true },
    ];

    const target = renderer.create(
      <Select
        label="Select something"
        title="Show sumthing"
        values={options}
        id="txt"
        value="Dos"
        handleChange={jest.fn()}
      />
    );

    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
