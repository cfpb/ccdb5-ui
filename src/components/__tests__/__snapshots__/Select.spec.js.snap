// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:Select renders array of objects without crashing 1`] = `
<section
  className="cf-select"
  data-tour="select-txt"
>
  <label
    className="u-visually-hidden"
    htmlFor="select-txt"
  >
    Select something
  </label>
  <p>
    Show sumthing
  </p>
  <select
    id="select-txt"
    onChange={[MockFunction]}
    value="Dos"
  >
    <option
      value="Uno"
    >
      Uno
    </option>
    <option
      value="Dos"
    >
      Dos
    </option>
    <option
      value="Tres"
    >
      Tres
    </option>
  </select>
</section>
`;

exports[`component:Select renders array values without crashing 1`] = `
<section
  className="cf-select"
  data-tour="select-txt"
>
  <label
    className="u-visually-hidden"
    htmlFor="select-txt"
  >
    Select something
  </label>
  <p>
    Show sumthing
  </p>
  <select
    id="select-txt"
    onChange={[MockFunction]}
    value="Dos"
  >
    <option
      disabled={false}
      value="Uno"
    >
      Uno
    </option>
    <option
      disabled={false}
      value="Dos"
    >
      Dos
    </option>
    <option
      disabled={false}
      value="Tres"
    >
      Tres
    </option>
  </select>
</section>
`;

exports[`component:Select renders object values without crashing 1`] = `
<section
  className="cf-select"
  data-tour="select-size"
>
  <label
    className="u-visually-hidden"
    htmlFor="select-size"
  >
    Select size
  </label>
  <p>
    Show
  </p>
  <select
    id="select-size"
    onChange={[MockFunction]}
    value={10}
  >
    <option
      disabled={false}
      value="10"
    >
      10 results
    </option>
    <option
      disabled={false}
      value="25"
    >
      25 results
    </option>
    <option
      disabled={false}
      value="50"
    >
      50 results
    </option>
    <option
      disabled={false}
      value="100"
    >
      100 results
    </option>
  </select>
</section>
`;
